File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Effect/Combat/IntenseZoomParticle.ts
import { FairyGUI } from "csharp";
import { Mathf } from "../../../../../Core/Module/Math/Mathf";
import { TimeKit } from "../../../../../Core/Utils/TimeKit";
import Color from "../../DataDefine/Color";
import FileHelper from "../../FileHelper";
import UIHelper from "../../UI/UIHelper";
import { AbstractEffect } from "../AbstractEffect";

export class IntenseZoomParticle extends AbstractEffect{
    public Loader:FairyGUI.GLoader;
    private _FlickerDuration:number;
    private constructor(private _X,private _Y,private _IsBlack) {
        super();
        let i = Mathf.RandomRange(0,3);
        this.Loader = UIHelper.CreateGLoader()
        this.Loader.url = FileHelper.FormatPath("Vfx/cone" + (i == 0 ? "8" :(i == 1 ? "5" : "6"))); 
        this.Loader.SetPosition(_X,_Y,0);
        this.Loader.color = _IsBlack ? Color.Black.UnityColor : Color.Gold.UnityColor;
        if(!_IsBlack)
            this.Loader.blendMode = FairyGUI.BlendMode.Add;
        this.Randomize();
    }
    
    Pow2Out(a:number) : number{
        return Math.pow(a - 1.0, 2) * (2 % 2 == 0 ? -1 : 1) + 1.0;
    }
    
    Randomize(){
        this.Loader.rotation = Mathf.RandomRange(0,360);
        let offsetX = Mathf.RandomRange(200,600) * (2 - (1.5 - this.Duration));
        this.Loader.SetPosition(this._X + offsetX,this._Y,0);
        this.Loader.SetPivot(this.Loader.texture.width / offsetX,0.5);
        if(this._IsBlack)
            this.Loader.alpha = Mathf.RandomRange(0.5,1) * this.Pow2Out(1.5-this.Duration / 1.5);
        else
            this.Loader.alpha = Mathf.RandomRange(0.2,0.7) * this.Pow2Out(1.5-this.Duration / 1.5);
    }
    
    Update() {
        super.Update();
        this._FlickerDuration -= TimeKit.DeltaTime;
        if(this._FlickerDuration < 0){
            this.Randomize();
            this._FlickerDuration = Mathf.RandomRange(0,0.05);
        }
        if(this.Duration >= 1.5)
            this.IsDone = true;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Effect/Combat/IntenseZoomParticle.cs
using FairyGUI;
using ModTheSpire.DataDefine;
using UnityEngine;

namespace ModTheSpire.Effect.Combat
{
    public class IntenseZoomParticle : AbstractEffect
    {
        private float mX;
        private float flickerDuration = 0.0f;
        private GLoader mLoader;
        private float mAlpha;
        public IntenseZoomParticle(float x, float y, bool isBlack)
        {
            mX = x;
            int i = Random.Range(0,2);
            mLoader = UIObjectFactory.NewObject(ObjectType.Loader).asLoader;
            mLoader.autoSize = true;
            mLoader.url = i == 0 ? ImageMaster.CONE_2 : (i == 1 ? ImageMaster.CONE_4 : ImageMaster.CONE_5); 
            mLoader.color = isBlack ? Color.black : MTSColor.Gold;
            mLoader.SetPivot(0, 0.5f, true);
            mLoader.SetPosition(x,y,0);
            GTween.To(1f, 0f, 1.5f)
                .OnUpdate(t1 => mAlpha = (float)(isBlack ? Random.Range(0.5f, 1.0f) * t1.value.d : Random.Range(0.2f, 0.7f) * t1.value.d))
                .SetEase(EaseType.QuadOut).SetTarget(this);
            this.randomize();
        }

        public void randomize() {
            mLoader.rotation = Random.Range(0,360.0f);
            mLoader.x = mX + Random.Range(200.0f, 600.0f) * this.Duration;
            mLoader.scaleX = Random.Range(1.0f, 1.3f);
            mLoader.scaleY = Random.Range(0.9f, 1.2f);
            mLoader.alpha = mAlpha;
        }
        
        public override void Update()
        {
            base.Update();
            this.flickerDuration -= Time.deltaTime;
            if (this.flickerDuration < 0.0f) {
                this.randomize();
                this.flickerDuration = Random.Range(0.0f, 0.05f);
            }
            if (Duration >= 1.5f)
            {
                Finish();
            }
        }

        public override void Dispose()
        {
            base.Dispose();
            mLoader.Dispose();
            GTween.Kill(this);
        }
    }
}